sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,711 @@
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A IO to publish or consume messages with a RabbitMQ broker.
+ *
+ * <h3>Consuming messages from RabbitMQ server</h3>
+ *
+ * <p>{@link RabbitMqIO} {@link Read} returns an unbounded {@link PCollection} containing RabbitMQ
+ * messages body (as {@code byte[]}).
+ *
+ * <p>To configure a RabbitMQ source, you have to provide a RabbitMQ {@code URI} to connect
+ * to a RabbitMQ broker. The following example illustrates various options for configuring the
+ * source:
+ *
+ * <pre>{@code
+ *
+ *  pipeline.apply(
+ *    RabbitMqIO.read().withUri("amqp://user:password@localhost:5672").withQueue("QUEUE")
+ *
+ * }</pre>
+ *
+ * <h3>Publishing messages to RabbitMQ server</h3>
+ *
+ * <p>{@link RabbitMqIO} {@link Write} can send {@code byte[]} to a RabbitMQ server queue.
+ *
+ * <p>As for the {@link Read}, the {@link Write} is configured with a RabbitMQ
+ * {@link ConnectionConfig}.
+ *
+ * <p>For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...) // provide PCollection<byte[]>
+ *    .apply(RabbitMqIO.write().withUri("amqp://user:password@localhost:5672").withQueue("QUEUE"));
+ *
+ * }</pre>
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class RabbitMqIO {
+
+  public static Read read() {
+    return new AutoValue_RabbitMqIO_Read.Builder()
+        .setConnectionConfig(ConnectionConfig.create()).setQueueDeclare(false)
+        .setMaxReadTime(null).setMaxNumRecords(Long.MAX_VALUE).setUseCorrelationId(false).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_RabbitMqIO_Write.Builder()
+        .setConnectionConfig(ConnectionConfig.create())
+        .setExchangeDeclare(false).build();
+  }
+
+  private RabbitMqIO() {
+  }
+
+  /**
+   * Describe a connection configuration to a RabbitMQ server.
+   */
+  @AutoValue
+  public abstract static class ConnectionConfig implements Serializable {
+
+    @Nullable abstract String uri();
+
+    abstract int networkRecoveryInterval();
+    abstract boolean automaticRecovery();
+    abstract boolean topologyRecovery();
+
+    abstract int connectionTimeout();
+    abstract int requestedChannelMax();
+    abstract int requestedFrameMax();
+    abstract int requestedHeartbeat();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setUri(String uri);
+      abstract Builder setNetworkRecoveryInterval(int networkRecoveryInterval);
```

Comment:
Same comments about these parameters as in the JDBC PR. Do not mirror every parameter of the underlying library - expose only parameters that are absolutely essential, and have predictable impact regardless of the implementation of RabbitMqIO or the runner behavior.